### ISSUE TITLE
[scoped-elements] Detect & act on number of implementations 

### ISSUE THREAD

**Issue by jorenbroekema** [open-wc/open-wc#1369]
Hey,

Feature request here.

As far as I understand it is possible for the ScopedElementsMixin to detect how often a certain component tries to register on the custom element registry through scopedElements getter, so it could count for e.g. `lion-feedback` scoped element, how often it registers. In addition, on the first time, we could even test if it's already registered under unscoped lion-feedback (e.g. from an old version that doesn't use the ScopedElementsMixin).

That way we can count the amount of implementations for a specific element, for example for an app using three features that rely on v1 of lion-feedback, v2, and v3 respectively. 

### Why is that useful?

Because now we can error out, or warn, through a flag e.g. `maxImplementationCount`, for example if set to 1, it would error on runtime if you have more than 1 `lion-feedback`, regardless of whether it is unscoped or scoped. This is something that may help people troubleshoot a bit (and guard for) more how many versions of the same element are being registered, by which of their dependencies this is caused, etc.

### Flow

- lion-feedback (v1) can be registered normally through explicit side effect file that does `customElements.define('lion-feedback', LionFeedback)`. 
- LionFeedback (v2) class gets imported and scoped through ScopedElementsMixin as lion-feedback-12345, but the mixin also sees that lion-feedback already exists as it is on the registry --> implementation-count for this component = 2
- LionFeedback (v3) class gets imported and scoped as lion-feedback-45678 --> implementation-count for this component = 3

Problem I see here is: what if lion-feedback (v1) file gets loaded last, then the mixin doesn't know about it and the implementation-count would be 1 lower than actual count? Could the mixin somehow listen to the CustomElementRegistry registrations for names being added that are also existing keys in the mixin?

**Comment by LarsDenBakker** [comment]
I think we can implement this as a decorator on the global customElementa.define, that also prevents race conditions.

**Comment by frank-dspeed** [comment]
@jorenbroekema since you got javascript code so it is text you should be able to do a simple search on your file to know how often something is registered and if you got a central method to register your components there is something wrong already.

**Comment by jorenbroekema** [comment]
@frank-dspeed Not just my code, can also be dependencies. I prefer Lars' idea since it is far less hacky, far less error prone, and works on runtime rather than as an external script or build-step plugin

**Comment by frank-dspeed** [comment]
it is ok for me I only needed to drop that so when someone reads this he will understand that this is only for developer joy not for quality or anything other. 

And I need to say hacky is to replace names on runtime. Your Code should be in a condition that works without hacking any registry stuff together ;)

**Comment by jorenbroekema** [comment]
There are some contexts in my company at least where this feature goes beyond just developer joy but rather a necessity in governing the amount of versions of the same component are being registered.

And yes it is hacky but until scoped elements spec lands in the browser, we have to deal with the cards we are handed

**Comment by frank-dspeed** [comment]
there is no scoped element spec for the browser we discussed that and agreed on no need for that as you can scope everything easy your self there are 100 of patterns to archive that.

**Comment by jorenbroekema** [comment]
So please correct me if I misunderstand, are you saying "Just use static analysis in your build-step" as a solution to this issue? Because if so, thank you but no, that's not really meeting my requirements, because I'd like to see something working on runtime.

**Comment by frank-dspeed** [comment]
@jorenbroekema yes i would say static handling that is the only way to go and this scoped element hack thing is for people that can not code themself but need to use 3th party components that are self-registering.

This is really for edge cases imagine the overhead and lucky the ECMAScript Spec evaluates a module only once so if you import a self-registering component that does that as a side effect that will get executed only once in the browser since that has a global module registry even if you import the module in different places

**Comment by frank-dspeed** [comment]
Module that self-registering-component.js
```js
customeElements.define()
``` 

a.js
``` 
import 'self-registering-component.js'
``` 

b.js
``` 
import 'self-registering-component.js'
``` 

c.js
``` 
import 'self-registering-component.js'
``` 

customElements.define gets executed exactly once!

**Comment by jorenbroekema** [comment]
It would get executed twice on the same custom element name if you have a dependency that uses v1 of a certain component (let's say a third party tooltip component) and you yourself are using v2 of this same component.

I know how ES Modules work.... this is not an ES Modules problem. This is a problem of having multiple major versions of the same component end up in the same application. I would advise you to read https://open-wc.org/scoped-elements/ . If you still think there is no reason for scoped elements, let's discuss that in a separate issue, and not derail this one.

**Comment by frank-dspeed** [comment]
@jorenbroekema ah ok sure that's also a legit case for this.  i agree but still counting how often something registers does not help any way I am out :) 

but for the version management, I would choose npm ls component-name but i now fully understand your case you need it as a dev tool :)
